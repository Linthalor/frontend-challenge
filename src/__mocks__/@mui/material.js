const genMockComponent = (name) => (props) => {
  const MockName = `mocked-mui-${name}`;
  const {sx, ...remainingProps} = props;
  return <MockName {...remainingProps} sx={sx ? JSON.stringify(sx) : undefined} />;
}

module.exports = {
  ...jest.requireActual('@mui/material'),
  Box: genMockComponent('box'),
  Button: genMockComponent('button'),
  IconButton: genMockComponent('icon-button'),
  InputAdornment: genMockComponent('input-adornment'),
  TextField: genMockComponent('text-field'),
  Typography: genMockComponent('typography'),
  Card: genMockComponent('card'),
  CardContent: genMockComponent('card-content'),
  CardActions: genMockComponent('card-actions'),
};