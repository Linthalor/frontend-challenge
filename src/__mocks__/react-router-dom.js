module.exports = {
  ...jest.requireActual('react-router-dom'),
  Navigate: (props) => {
    const MockName = `mocked-react-router-dom-navigate`;
    return <MockName {...props} />;
  }
};