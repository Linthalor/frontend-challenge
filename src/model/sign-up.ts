import { z } from 'zod';
import { t } from 'i18next';

const minNameSize = 1;
const nameMessages = (t('form:name:errors', { returnObjects: true }) as any)(minNameSize);

const minPasswordSize = 8;
const passwordMessages = (t('form:name:errors', { returnObjects: true }) as any)(minPasswordSize);

export const SignUpUserFormValuesSchema = z.object({
  name: z.string({
    required_error: nameMessages.required_error,
  }).min(minNameSize, nameMessages.too_small),
  email: z.string({
    required_error: t('form:email:errors:required_error'),
  }).email(t('form:email:errors:invalid_string')),
  password: z.string({
    required_error: passwordMessages.required_error,
  }).min(minPasswordSize, passwordMessages.too_small),
});
export type SignUpUserFormValues = z.infer<typeof SignUpUserFormValuesSchema>;

// Used to effectively prevent an empty selection.
const minColorSize = 1;

export const SignUpUserPrefsValuesSchema = z.object({
  // TODO: have this fetch the redux's value to compare for validation? might need a component to do that...
  color: z.string({
    required_error: t('form:color:errors:required_error'),
  }).min(minColorSize),
  terms: z.boolean().refine(val => val, {}),
});
export type SignUpUserPrefsValues = z.infer<typeof SignUpUserPrefsValuesSchema>;

export type SignUp = SignUpUserFormValues & SignUpUserPrefsValues;