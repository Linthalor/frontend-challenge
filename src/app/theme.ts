import { createTheme } from '@mui/material';

// Themed using Upgrade's palette
export const theme = createTheme({
  palette: {
    primary: {
      light: '#60c7a1',
      main: '#00a669',
      dark: '#008051',
    },
    secondary: {
      main: '#3f3d56',
      light: '#6bc1fe',
    },
    success: {
      main: '#03b57a',
      contrastText: '#ffffff',
    },
    error: {
      main: '#c83c4c',
      contrastText: '#ffffff',
    },
    text: {
      primary: '#3f465e',
      secondary: '#008051'
    }
  },
});