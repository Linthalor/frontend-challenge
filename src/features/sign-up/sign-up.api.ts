import { SignUp } from '../../model/sign-up';
import { appPost, baseUrl } from '../api';

/*
  I am ignoring the fact that the password will be sent in plain text across an unsecure
  connection as the mock backend is not using https etc. (Assuming intended simplicity).
*/
export const postSignUp = (
  SignUpData: SignUp
): Promise<true> => appPost(`${baseUrl}submit`, SignUpData).then(() => true);