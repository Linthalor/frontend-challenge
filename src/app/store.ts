import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { signUpReducer } from '../features/sign-up/sign-up.slice';
import { colorReducer } from '../features/color/color.slice';

export const store = configureStore({
  reducer: {
    signUp: signUpReducer,
    color: colorReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
