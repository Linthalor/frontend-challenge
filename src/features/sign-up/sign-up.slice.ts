import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../app/store';
import { SignUp, SignUpUserFormValues, SignUpUserPrefsValues } from '../../model/sign-up';
import { AsyncProp, initialAsync } from '../../model/util/async-prop';
import { postSignUp } from './sign-up.api';

export interface AuthState {
  signUpStatus: AsyncProp<true>,
  userFormValues: SignUpUserFormValues | null,
  userPrefsValues: SignUpUserPrefsValues | null,
};

const initialState = (): AuthState => ({
  signUpStatus: initialAsync(),
  userFormValues: null,
  userPrefsValues: null,
});

export const singUpAsync = createAsyncThunk<true, SignUp, { state: RootState }>(
  'auth/postSignUp',
  async (SignUpData) => await postSignUp(SignUpData),
);

export const authSlice = createSlice({
  name: 'auth',
  initialState: initialState(),
  reducers: {
    resetSignUp: () => initialState(),
    setUserFormValues: (state, action: PayloadAction<SignUpUserFormValues>) => {
      state.userFormValues = action.payload;
    },
    setUserPrefsValues: (state, action: PayloadAction<SignUpUserPrefsValues>) => {
      state.userPrefsValues = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(singUpAsync.pending, (state) => {
        state.signUpStatus = {
          loading: true,
        };
      })
      .addCase(singUpAsync.fulfilled, (state, action) => {
        state.signUpStatus = {
          value: action.payload,
        };
      })
      .addCase(singUpAsync.rejected, (state, action) => {
        state.signUpStatus = {
          errorMessage: action.error.message || 'Failed!'
        };
      });
  }
})

export const { resetSignUp, setUserFormValues, setUserPrefsValues } = authSlice.actions;

export const selectSignUpStatus = (state: RootState) => state.auth.signUpStatus;
export const selectUserFormValues = (state: RootState) => state.auth.userFormValues;
export const selectUserPrefsValues = (state: RootState) => state.auth.userPrefsValues;

export const authReducer = authSlice.reducer;