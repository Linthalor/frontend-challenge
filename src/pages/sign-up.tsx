import { Box, Button, IconButton, InputAdornment, TextField, Typography } from '@mui/material';
import { SignUpCard } from '../components/signup-card';
import { SignUpUserFormValues, SignUpUserFormValuesSchema } from '../model/sign-up';
import { useAppDispatch } from '../app/hooks';
import { selectUserFormValues, setUserFormValues } from '../features/sign-up/sign-up.slice';
import { useNavigate } from 'react-router-dom';
import {zodResolver} from '@hookform/resolvers/zod';
import { useSelector } from 'react-redux';
import { EmailRounded, Visibility, VisibilityOff } from '@mui/icons-material';
import { useForm } from 'react-hook-form';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';

export const SignUp = () => {
  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const { t } = useTranslation();
  const signUpUserData = useSelector(selectUserFormValues);
  const [seePassword, setSeePassword] = useState(false);

  const { handleSubmit, register, formState } = useForm({
    resolver: zodResolver(SignUpUserFormValuesSchema),
    mode: 'onBlur',
    defaultValues: {
      name: signUpUserData?.name || '',
      email: signUpUserData?.email || '',
      password: signUpUserData?.password || '',
    },
  });
  const { errors } = formState;

  const submitForm = (values: SignUpUserFormValues) => {
    dispatch(setUserFormValues(values));
    navigate('/more-info');
  }

  return (
    <form
      style={{
        height: '100%',
        width: '100%',
      }}
      onSubmit={handleSubmit(submitForm)}
    >
      <SignUpCard
        content={<>
          <Typography
            sx={{
              textAlign: 'center',
              fontWeight: 'bold'
            }}
            variant="h5"
            color="text.primary"
            gutterBottom
          >
            {t('page:signUp:label')}
          </Typography>
          <Box sx={{ gap: 2, display: 'flex', flexDirection: 'column' }}>
            <TextField
              id="first-name-field"
              label={t('form:name:label')}
              variant="outlined"
              type="text"
              required
              {...register('name')}
              error={!!errors.name?.message}
              helperText={errors.name?.message}
            />
            <TextField
              id="email-field"
              label={t('form:email:label')}
              variant="outlined"
              type="email"
              required
              {...register('email')}
              error={!!errors.email?.message}
              helperText={errors.email?.message}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <IconButton disabled>
                      <EmailRounded />
                    </IconButton>
                  </InputAdornment>
                ),
              }}
            />
            <TextField
              id="password-field"
              label={t('form:password:label')}
              variant="outlined"
              type={seePassword ? 'text' : 'password'}
              required
              {...register('password')}
              error={!!errors.password?.message}
              helperText={errors.password?.message}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <IconButton onClick={() => setSeePassword(!seePassword)}>
                      {seePassword ? <VisibilityOff /> : <Visibility />}
                    </IconButton>
                  </InputAdornment>
                ),
              }}
            />
          </Box>
        </>}
        actions={<>
          <Button variant="contained" type="submit" disabled={!formState.isValid}>
            {t('common:next')}
          </Button>
        </>}
      />
    </form>
  );
};