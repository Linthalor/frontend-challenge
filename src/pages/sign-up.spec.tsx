import { cleanup, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createTestStore, store as realStore } from '../app/store';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from '@mui/material';
import { theme } from '../app/theme';
import { SignUp } from './sign-up';

jest.mock('@mui/material');

describe('SignUp', () => {
  let store: typeof realStore;
  beforeEach(() => {
    store = createTestStore();
  });
  afterEach(() => {
    cleanup();
  });

  it('renders as expected', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <BrowserRouter>
            <SignUp />
          </BrowserRouter>
        </Provider>
      </ThemeProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});