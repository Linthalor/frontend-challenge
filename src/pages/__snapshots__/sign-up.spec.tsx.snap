// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sign-up page renders as expected 1`] = `
<form
  style="height: 100%; width: 100%;"
>
  <div
    class="MuiBox-root css-1jpcp03"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1vfmm4m-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root css-jewu2s-MuiCardContent-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-9vlyci-MuiTypography-root"
        >
          Sign Up
        </h5>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-zq33gc-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="first-name-field"
            id="first-name-field-label"
          >
            First Name
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-1o0vv86-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1kka5ew-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby="first-name-field-helper-text"
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="first-name-field"
              required=""
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  First Name
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-12kfrt5-MuiFormHelperText-root"
            id="first-name-field-helper-text"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </p>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-zq33gc-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="email-field"
            id="email-field-label"
          >
            Email
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-1o0vv86-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-1gjvb0p-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby="email-field-helper-text"
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="email-field"
              required=""
              type="email"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EmailRoundedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm-.4 4.25-7.07 4.42c-.32.2-.74.2-1.06 0L4.4 8.25c-.25-.16-.4-.43-.4-.72 0-.67.73-1.07 1.3-.72L12 11l6.7-4.19c.57-.35 1.3.05 1.3.72 0 .29-.15.56-.4.72z"
                />
              </svg>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Email
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-12kfrt5-MuiFormHelperText-root"
            id="email-field-helper-text"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </p>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-zq33gc-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="password-field"
            id="password-field-label"
          >
            Password
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-1o0vv86-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-1gjvb0p-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby="password-field-helper-text"
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="password-field"
              required=""
              type="password"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="LockRoundedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zM9 8V6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9z"
                />
              </svg>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Password
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-12kfrt5-MuiFormHelperText-root"
            id="password-field-helper-text"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </p>
        </div>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1lljpoj-MuiCardActions-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-1ppzino-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="submit"
        >
          Next
        </button>
      </div>
    </div>
  </div>
</form>
`;
